Add denominator adjusted SLI type

gbbirkisson opened the pull request:
## The Problem

The regular events SLI type does well when you have abundant amount of events that are distributed evenly across each hour of the day. That SLI type starts to exhibit some problems when we deviate from that criteria.

As an example, if we have an hour that has an error-ratio of $1 \over 100000$, and the next hour has an error-ratio of $1 \over 2$, the latter hour will have a greater impact on the burn-rate and error budget. We call this the denominator problem.

If you have a service that has little traffic during the night, you will likely encounter this problem in your error budget.

## The Solution

This PR address this problem by introducing a new SLI type called `denominator_corrected`. That SLI adds recording rules for storing how the current number of events compares to the average of the preceding month. This allows an SLO to burn faster or slower depending on whether there are a lot or few events in the denominator. This will make sure that SLO's burn slower when there are few events. 

To make this new SLI flexible, we allow users either to supply `error_query` or `success_query`, as well as `total_query` to calculate their error ratio.

We define $total_w$ as the `total_query` over a specific window $w$ and $total_m$ as the `total_query` over the last 30 days. Then we can define correction factor over the window $w$ as:

$$c_w = {total_w \over total_m}$$

If user supplies `error_query`, the error ratio for window $w$ will be:

$$r_w = c_w ({error_w \over total_w})$$

Else if user supplies `success_query`, the error ratio for window $w$ will be:

$$r_w = c_w (1 - {success_w \over total_w})$$


## Caveats

Although the the budget will burn slower during low-event intervals, it will also burn the budget faster during high-event intervals. If you have high traffic during the day, and low traffic during the night, your budget will burn faster during the day, compared to using the regular event SLI type.

The SLI takes into account the last month to adjust the denominator. This can be problematic if your service sees big differences between months in traffic. If traffic last month was way higher then this month, this months budget will burn slower. Similarly if last month had low traffic, this month budget will burn faster.

Another side effect of adding essentially weights to each hour, is that in some cases the burn-rate can exceed 100%. This should not be a problem, but it is important to keep in mind.

r3code commented:
@RmStorm What did you correct in Sloth Dashboards? Can you post a link to repo with it? 
It looks so that it requires some changes to support slo:numerator_correction:ratio{{.window}}

r3code commented:
@gbbirkisson @RmStorm Need some clarification. Can you please look at my example below.

_For example, we have:_  
* 90 000 requests as total for last 30d. It's 3000 req/day and 125 req/h in AVG. 
* Error budget 5% (SLO = 95%), 30 day window. So we have max_hours_with_errors = 0,05 * 30 * 24 = 0,05 * 720 = 0,5*72 = 36h   
AVG error rate = 36h/720h = 0,05  

_Assume we see:_  
* Day - `error count 1h [10:00-11:00] = 100, total = 120 -> err_rate_1h_day = 100/120 = 10/12 = 0,833, hours_with_errors = interval * err_rate = 1h * 0,833 =  0,833h, error_budget_burned = hours_with_errors / max_hours_with_errors = 0,833h/36h = 0,023 (2,3% in 1 hour) `
* Night - `error count 1h [01:00-02:00] = 10, total 13 -> err_rate_1h_night = 10/13 = 0,769, hours_with_errors = interval * err_rate = 1h * 0,769 =  0,769h, error_budget_burned = hours_with_errors / max_hours_with_errors = 0,769h/36h = 0,021 (2,1% in 1 hour)  `
Day to night `budget burnt ratio = 0,023/0,021 =~ 1,09`

_Correction factors are:_  
`c_1h_day = 120/90 000 = 12/9000 = 0,001333`  
`c_1h_night = 13/90 000 = 1/9000  = 0,000144`  

_With correction factors:_  
* Day - `err_rate_1h_day_corr = c_1h_day * err_rate_1h_day = 0,001333 *  0,833 = 0,00111, hours_with_errors = 1h * 0,00111 = 0,00111h, error_budget_burned = 0,00111/36 = 0,0000308333  (0,00308333%)`
* Night - `err_rate_1h_night_corr = c_1h_night  * err_rate_1h_night = 0,000144 * 0,769 = 0,00011, hours_with_errors = 1h * 0,00011 = 0,00011, error_budget_burned = 0,00011/36 = 0,0000030556 (0,00030556%)  `
Day to night `budget burnt ratio = 0,00308333/0,00030556 =~ 10,09`

Looks good it helps to burn less budget at night, but, I think I missed something in a math because my `error_budget_burned` values with corrections too low (681 times lower than without corrections). It will take too long to burn the whole budget.

gbbirkisson commented:
> Need some clarification. Can you please look at my example below.

Maybe the description on the issue is not quite clear enough. The correction factor is a number that hovers around 1. It takes a dip during the nights, and goes a bit over 1 during the day (assuming you have more events during the day). This graph shows the correction factor calculations in one of the tests we are running:

![image](https://github.com/slok/sloth/assets/6891726/31fa1969-d536-4e8a-baee-602a70e627d4)

So the math for your example would be something like:

Seconds in 1h: `60 * 60` = `3600`
Seconds in 30d: `3600 * 24 * 30` = `2592000`
Requests for last 30d: `90000`
Request rate last 30d (`total_m`): `90000 / 2592000` = `0.03472`
Average requests per hour (`avg_h`): `125`
Max hours with error: `36`

**10:00 - 11:00:**
Requests failed 1h: `100`
Request total 1h: `120`
Error rate (`error_w`): `100 / 3600` = `0.02777` 
Request rate (`total_w`): `120 / 3600` = `0.03333`
Error ratio: `(100 / 3600) / (120 / 3600)` = `100 / 120` = `0.83333`
Burn of budget: `0.83333 / 36` = `0.023` = `2.3%`

Correction factor (`c_w`): `0.03333 / 0.03472` = `0.95996` (= `total_w / avg_h` = `120 / 125`)
Corrected error ratio (`r_w`): `0.83333 * 0.95996` = `0.79999`
Corrected burn of budget: `0.79999 / 36` = `0.022` = `2.2%`

Because this particular hour has fewer requests than the average hour the last 30d it weighs less (correction factor below 1). By this logic you can imagine that hours with more than 125 requests will weigh more (correction factor above 1).

r3code commented:
@gbbirkisson thanks for your clarification! Now it's really clear to me. I faced the same problem, and I'm going to use your solution. 

P.S. it looks like @slok abandoned this project. He haven't been merging PRs since 22 Nov 2022 :(

gbbirkisson commented:
Great to hear that this solution makes sense @r3code. Feel free to use our fork: https://github.com/kolonialno/sloth

ishantanu commented:
Hello, regarding the support for `success_query`, was it added only for the new denominator adjusted SLI type or for all SLI types?

RmStorm commented:
> Hello @gbbirkisson , regarding the support for `success_query`, was it added only for the new denominator adjusted SLI type or for all SLI types?

Only for the denominator adjusted type [by the looks of it](https://github.com/kolonialno/sloth/blob/main/internal/prometheus/model.go#L18-L37)!
